repair/repair.hh

```diff
@@ -46,6 +46,8 @@ public:
     repair_stopped_exception() : repair_exception("Repair stopped") { }
 };
 
+future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_metadata tm, std::unordered_set<dht::token> tokens);
```

Comment:
Please add a proper description of the function. In particular what `tokens` are here?

Reply:
The tokens here are tokens the node selected to bootstrap the node.

Reply:
Fixed.

Reply:
> The tokens here are tokens the node selected to bootstrap the node.

Why do we need to pass them? Can't we simply get them from the gossiper/topology info? The same place where you get them when you call `get_saved_tokens()`...

I don't see that you persist those tokens in the `bootstrap_with_repair()`.
If that's indeed the case you may want to minimize the number of different "copies" of this read-only data structure if you can do that efficiently.

Reply:
Because we do not want to add unnecessary dependencies between modules. 
The tokens are persisted before we call bootstrap() in storage_service.cc. 
The tokens object is 256 tokens. I do not see a problem or the need to avoid copying it.

Reply:
My concern is not about copying 256 tokens but rather the lack of that dependency that you seem to dislike. 

If you copy specific data around you may find yourself having to make sure that the data copies are in sync as some point. At that point having that dependency adds you an extra guarantee that the data view in this and another component are the same.

Ideally, everybody should access the read-only data by reference under read_lock.

Reply:
The upper layer, storage_service, chooses the tokens for the node. It asks repair service, bootstrap_with_repair(), to bootstrap. The tokens are not going to change by design.